plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+            lock(lock),
+            lock_surface(lock_surface),
+            output(output),
+            interaction(std::make_unique<lock_surface_keyboard_interaction>(lock_surface->surface,
+                lock_surface->output))
+        {
+            lock_surface_destroy.set_callback([this] (void *data)
+            {
+                wf::wlr_surface_controller_t::try_free_controller(this->lock_surface->surface);
+                wf::scene::remove_child(shared_from_this());
+                lock_surface_destroy.disconnect();
+                this->lock->surface_destroyed(this->output);
+                const char *name = this->output->handle ? this->output->handle->name : "(deleted)";
+                LOGD("lock_surface on ", name, " destroyed");
+            });
+            lock_surface_destroy.connect(&lock_surface->events.destroy);
+        }
+
+        void attach_to_layer()
+        {
+            auto layer_node = output->node_for_layer(wf::scene::layer::LOCK);
+            wf::scene::add_front(layer_node, shared_from_this());
+            wf::wlr_surface_controller_t::create_controller(lock_surface->surface, layer_node);
+            focus();
+        }
+
+        void focus()
+        {
+            wf::get_core().seat->set_active_node(shared_from_this());
+            interaction->handle_keyboard_enter(wf::get_core().seat.get());
+        }
+
+        wf::keyboard_focus_node_t keyboard_refocus(wf::output_t *output) override
+        {
+            wf::keyboard_focus_node_t node = {
+                .node = this,
+                .importance = wf::focus_importance::HIGH,
+                .allow_focus_below = false,
+            };
+            return node;
+        }
+
+        class lock_surface_keyboard_interaction : public wf::keyboard_interaction_t
+        {
+          public:
+            lock_surface_keyboard_interaction(wlr_surface *surface, wlr_output *output) : surface(surface),
+                output(output)
+            {}
+
+            void handle_keyboard_enter(wf::seat_t *seat)
+            {
+                wlr_seat_keyboard_enter(seat->seat, surface, nullptr, 0, nullptr);
+                LOGI("Focused ", output->name);
+            }
+
+            void handle_keyboard_leave(wf::seat_t *seat)
+            {
+                wlr_seat_keyboard_clear_focus(seat->seat);
+                LOGI("Unfocused ", output->name);
+            }
+
+            void handle_keyboard_key(wf::seat_t *seat, wlr_keyboard_key_event event)
+            {
+                wlr_seat_keyboard_notify_key(seat->seat, event.time_msec, event.keycode, event.state);
+            }
+
+          private:
+            wlr_surface *surface;
+            wlr_output *output;
+        };
+
+        wf::keyboard_interaction_t& keyboard_interaction()
+        {
+            return *interaction;
+        }
+
+      private:
+        wayfire_session_lock *lock;
+        wlr_session_lock_surface_v1 *lock_surface;
+        wf::output_t *output;
```

Comment:
Re lifetimes; I see that the output is a necessary thing to keep a reference to, which means you ought to also have a handler for when it is destroyed.